crates/wysiwyg/src/tests/test_to_markdown.rs

```diff
@@ -0,0 +1,228 @@
+// Copyright 2022 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{tests::testutils_composer_model::cm, ToMarkdown};
+use pulldown_cmark as md_parser;
+use widestring::Utf16String;
+
+#[test]
+fn text() {
+    assert_to_md("abc", "abc");
+    assert_to_md("abc def", "abc def");
+    // Internal spaces are preserved.
+    assert_to_md("abc   def", "abc   def");
+}
+
+#[test]
+fn text_with_linebreaks() {
+    // One new line.
+    assert_to_md(
+        "abc<br />def",
+        r#"abc\
+def"#,
+    );
+
+    // Two new lines (isn't transformed into a new block).
+    assert_to_md(
+        "abc<br /><br />def",
+        r#"abc\
+\
+def"#,
+    );
+}
+
+#[test]
+fn text_with_italic() {
+    assert_to_md("<em>abc</em>", "*abc*");
+    // Internal emphasis.
+    assert_to_md("abc<em>def</em>ghi", "abc*def*ghi");
+    assert_to_md("abc <em>def</em> ghi", "abc *def* ghi");
+    assert_to_md_no_roundtrip("abc<em> def </em>ghi", "abc* def *ghi");
+    assert_to_md(
+        "abc <em>line1<br />line2<br /><br />line3</em> def",
+        r#"abc *line1\
+line2\
+\
+line3* def"#,
+    );
+}
+
+#[test]
+fn text_with_bold() {
+    assert_to_md("<strong>abc</strong>", "__abc__");
+    assert_to_md_no_roundtrip("abc<strong>def</strong>ghi", "abc__def__ghi");
```

Comment:
Assuming this markdown is valid, we could consider fixing this as a bug in our markdown library (or choosing another that works better?)

Reply:
To clarify, the problem doesn't come from “our” code, but from `pulldown` I guess. The problem is that Markdown is not very clear about what to do here. I'll see with `pulldown` why they did this as such.

Reply:
Yeah, just musing here that we could consider reporting/fixing a bug in pulldown.